src/rnnlm/arpa-sampling.cc

```diff
@@ -0,0 +1,402 @@
+    word_pair = std::make_pair(word_s, word);
+    vocab_.push_back(word_pair);
+  }
+}
+
+void ArpaSampling::HeaderAvailable() {
+  ngram_counts_ = NgramCounts();
+  ngram_order_ = NgramCounts().size(); 
+  probs_.resize(ngram_order_);
+}
+
+BaseFloat ArpaSampling::GetProb(int32 order, int32 word, const HistType& history) {
+  BaseFloat prob = 0.0;
+  auto it = probs_[order - 1].find(history);
+  if (it != probs_[order - 1].end() &&
+      probs_[order-1][history].find(word) != probs_[order-1][history].end()) {
+    prob += probs_[order-1][history][word].first;
+  } else { // backoff to the previous order
+    order--;
+    if (order >= 1) {
+      HistType::const_iterator first = history.begin() + 1;
+      HistType::const_iterator last = history.end();
+      HistType h(first, last);
+      prob += GetProb(order, word, h);
+      int32 word_new = history.back();
+      HistType::const_iterator last_new = history.end() - 1;
+      HistType h_new(history.begin(), last_new);
+      prob += GetBackoffWeight(order, word_new, h_new);
+    }
+  }
+  return prob;
+}
+
+BaseFloat ArpaSampling::GetBackoffWeight(int32 order, int32 word, const HistType& history) {
+  BaseFloat bow = 0.0;
+  auto it = probs_[order - 1].find(history);
+  if (it != probs_[order - 1].end()) {
+    auto it2 = probs_[order - 1][history].find(word);
+    if (it2 != probs_[order - 1][history].end()) {
+      bow = (it2->second).second;
+    }
+  }
+  return bow;
+}
+
+void ArpaSampling::ComputeWordPdf(const HistType& history, std::vector<std::pair<int32, BaseFloat> >* pdf) {
+  int32 order = history.size();
+  BaseFloat prob = 0.0;
+  for (int32 i = 0; i < num_words_; i++) {
+    auto it = probs_[order].find(history);
+    int32 word = vocab_[i].second; // get word from the map
+    if (it != probs_[order].end()) {
+      auto it2 = probs_[order][history].find(word);
+      if (it2 != probs_[order][history].end()) {
+        prob = pow(10, (it2->second).first);
+        (*pdf)[i].first = word;
+        (*pdf)[i].second += prob;
+      } else {
+        HistType::const_iterator first = history.begin() + 1;
+        HistType::const_iterator last = history.end();
+        HistType h(first, last);
+        int32 word_new = history.back();
+        HistType::const_iterator last_new = history.end() - 1;
+        HistType h_new(history.begin(), last_new);
+        prob = pow(10, GetBackoffWeight(order, word_new, h_new)) *
+ 
+          pow(10, GetProb(order, word, h));
+        (*pdf)[i].first = word;
+        (*pdf)[i].second += prob;
+      }
+    } else {
+      HistType::const_iterator first = history.begin() + 1;
+      HistType::const_iterator last = history.end();
+      HistType h(first, last);
+      int32 word_new = history.back();
+      HistType::const_iterator last_new = history.end() - 1;
+      HistType h_new(history.begin(), last_new);
+      prob = pow(10, GetBackoffWeight(order, word_new, h_new)) *
+             pow(10, GetProb(order, word, h));
```

Comment:
if you write pow(10, a) * pow(10, b) you should replace it with
pow(10, a + b) which is more efficient